### PR TITLE
HIVE-24510: Vectorize compute_bit_vector

### PR DESCRIPTION
**Pull request by mustafaiman**
I organized this pull request in 3 commits for ease of review.
First commit adds a test that prints hex(compute_bit_vector(col)) for some tables. So we can use that as the source of truth for testing vectorized function.
Second commit is the actual implementation.
Third commit is the q.out file changes.

### CODE REVIEW

File: ql/src/test/results/clientpositive/llap/auto_sortmerge_join_14.q.out [hunk 1]
```diff
@@ -194,7 +222,7 @@ STAGE PLANS:
                       keys:
                         0 _col0 (type: int)
                         1 _col0 (type: int)
-                      Statistics: Num rows: 221 Data size: 1768 Basic stats: COMPLETE Column stats: COMPLETE
+                      Statistics: Num rows: 220 Data size: 1760 Basic stats: COMPLETE Column stats: COMPLETE
```
Comment (abstractdog) [inline ql/src/test/results/clientpositive/llap/auto_sortmerge_join_14.q.out:225]: have you double-checked the root cause of this one? I cannot decide at first sight whether this is a problem or not

Comment (mustafaiman) [reply ql/src/test/results/clientpositive/llap/auto_sortmerge_join_14.q.out:225]: Yes, before the patch distinct count on tbl2_n6 was miscalculated. Running the modified test before the patch reveals that distinct count on `key` was calculated wrong.
```
describe formatted tbl2_n6 key;
select count (distinct key) from tbl2_n6;
```


```
POSTHOOK: Input: default@tbl2_n6
col_name            	key                 
data_type           	int                 
min                 	0                   
max                 	199                 
num_nulls           	0                   
distinct_count      	117                 
avg_col_len         	                    
max_col_len         	                    
num_trues           	                    
num_falses          	                    
bit_vector          	HL                  
comment             	from deserializer   
COLUMN_STATS_ACCURATE	{\"BASIC_STATS\":\"true\",\"COLUMN_STATS\":{\"key\":\"true\",\"value\":\"true\"}}
...
POSTHOOK: query: select count (distinct key) from tbl2_n6
...
121
```

After the patch distinct count is correct. So the stats in the following query is correct now.

**Comment by abstractdog** [comment]
ran through the patches, I'll comment on particular code sections, but first a question in general: is there a specific reason for doing this method transformation?
```
compute_bit_vector(x, 'hll') -> compute_bit_vector_hll(x)
```
while I was reading the qout changes I realized that this rewrite makes this patch so huge, and I'm not sure if it's really worth, as I cannot see the added value, and I personally don't have any problems with the algorithm as a parameter, not the mention the possible fallback to a default maybe?
```
compute_bit_vector(x) -- could be translated to --> compute_bit_vector_hll(x, 'hll') 
```
I would consider this, as, without added value, e.g. the backport of this patch will be painful, and this patch is about vectorization coverage, and has nothing to do with compiler basically, what do you think?

File: ql/src/test/results/clientpositive/llap/dynpart_sort_optimization.q.out [hunk 2]
```diff
@@ -161,20 +168,28 @@ STAGE PLANS:
...
-                        key expressions: _col4 (type: tinyint)
-                        null sort order: a
-                        sort order: +
-                        Map-reduce partition columns: _col4 (type: tinyint)
-                        Statistics: Num rows: 1 Data size: 24 Basic stats: COMPLETE Column stats: NONE
-                        value expressions: _col0 (type: smallint), _col1 (type: int), _col2 (type: bigint), _col3 (type: float)
             Execution mode: llap
             LLAP IO: all inputs
         Reducer 2 
+            Execution mode: llap
```
Comment (abstractdog) [inline ql/src/test/results/clientpositive/llap/dynpart_sort_optimization.q.out:174]: could you please clarify why the patch has moved this operator from Reducer3 to Reducer2? basically, I "like" that we eliminated a reducer stage, however, I don't have an idea whether this is correct not, or what's the root cause behind it? 

as a reference, the history of dynamic partitioning sort optimization (what I can recall) is: HIVE-6455, HIVE-20703, HIVE-23071

Comment (mustafaiman) [reply ql/src/test/results/clientpositive/llap/dynpart_sort_optimization.q.out:174]: Actually we did not eliminate anything. If you look closely, what happened is Reducer 2 and Reducer 3 exchanged names. We still have the same number of mappers/reducers and the same operators overall. I am not sure why the names changed however they are arbitrary anyway. So I am not concerned with that.

**Comment by mustafaiman** [comment]
@abstractdog I had to introduce `compute_bit_vector_hll` and `compute_bit_vector_fm` because, vectorization is supported only for single argument functions. compute_bit_vector(arg1, arg2) will not vectorize.

File: ql/src/test/results/clientpositive/llap/dynpart_sort_optimization.q.out (hunk 2 above)
Comment (abstractdog) [reply ql/src/test/results/clientpositive/llap/dynpart_sort_optimization.q.out:174]: okay, thanks!

File: ql/src/test/results/clientpositive/llap/auto_sortmerge_join_14.q.out (hunk 1 above)
Comment (abstractdog) [reply ql/src/test/results/clientpositive/llap/auto_sortmerge_join_14.q.out:225]: cool, this is a nice catch then, thanks!

**Comment by abstractdog** [comment]
> @abstractdog I had to introduce `compute_bit_vector_hll` and `compute_bit_vector_fm` because, vectorization is supported only for single argument functions. compute_bit_vector(arg1, arg2) will not vectorize.

I think you can do it, refer to VectorUDAFBloomFilterMerge

it uses "matches" function without the param, which is ok, param value is handled later:
https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/exec/vector/expressions/aggregates/VectorUDAFBloomFilterMerge.java#L578

but then it's instantiated with parameters from conf:
https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByOperator.java#L1242

could this work for you?

**Comment by mustafaiman** [comment]
I made a quick fix to allow that in early versions of this patch. Then I decided to not pursue it because I did not see the need for allowing constant argument in runtime.

> you can still do something like:
> if compute_bit_vector: -> handle constant parameter
We do exactly that. Not in vectorizer but earlier in `ColumnStatsSemanticAnalyzer.java `. I am reluctant to implement extra functionality or add special cases unless it is necessary. Note that compute_bit_vector is a newly added UDF in 4.0. So there is no backward compatibility concern either.
Do you see any other benefit than preserving the earlier q.out outputs?

**Comment by abstractdog** [comment]
> I made a quick fix to allow that in early versions of this patch. Then I decided to not pursue it because I did not see the need for allowing constant argument in runtime.
> 
> > you can still do something like:
> > if compute_bit_vector: -> handle constant parameter
> 
> We do exactly that. Not in vectorizer but earlier in `ColumnStatsSemanticAnalyzer.java `. I am reluctant to implement extra functionality or add special cases unless it is necessary. Note that compute_bit_vector is a newly added UDF in 4.0. So there is no backward compatibility concern either.
> Do you see any other benefit than preserving the earlier q.out outputs?

no, I'm concerned only about the qout changes
you're right, if compute_bit_vector is a relatively new thing then we can also ignore backward compatibility problems and go on with compute_bit_vector_hll
I would personally keep pursuing a smaller patch as having "compute_bit_vector_hll" has no benefits either, but it's up to you, I think if the default hll algo won't be changed in the near future, we can go with the updated qouts :)

**Review by abstractdog — APPROVED** [review]
LGTM after comments and offline discussion
+1

**Pull request closed without merging**